rust/kernel/net/socket.rs

```diff
@@ -0,0 +1,641 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Socket API.
+//!
+//! This module contains the Socket layer kernel APIs that have been wrapped for usage by Rust code
+//! in the kernel.
+//!
+//! C header: [`include/linux/socket.h`](../../../../include/linux/socket.h)
+//!
+//! This API is inspired by the Rust std::net Socket API, but is not a direct port.
+//! The main difference is that the Rust std::net API is designed for user-space, while this API
+//! is designed for kernel-space.
+//! Rust net API: <https://doc.rust-lang.org/std/net/index.html>
+
+use super::*;
+use crate::error::{to_result, Result};
+use crate::net::addr::*;
+use crate::net::ip::IpProtocol;
+use crate::net::socket::opts::{OptionsLevel, WritableOption};
+use core::cmp::max;
+use core::marker::PhantomData;
+use flags::*;
+use kernel::net::socket::opts::SocketOption;
+
+pub mod flags;
+pub mod opts;
+
+/// The socket type.
+pub enum SockType {
```

Comment:
One of the most confusing things in the kernel has to be that `sock` and `socket` are different things. Unfortunately I don't have any better suggestions, but we probably want to make a more clear distinction on the Rust side.